data/types.go

```diff
@@ -238,21 +238,28 @@ type DelegatedRole struct {
 	Paths            []string `json:"paths"`
 }
 
-func (d *DelegatedRole) MatchesPath(file string) bool {
+// MatchesPath evaluates whether the path patterns or path hash prefixes match
+// a given file. This determines whether a delegated role is responsible for
+// signing and verifying the file.
+func (d *DelegatedRole) MatchesPath(file string) (bool, error) {
+	if err := d.validateFields(); err != nil {
+		return false, err
+	}
+
 	for _, pattern := range d.Paths {
 		if matched, _ := filepath.Match(pattern, file); matched {
-			return true
+			return true, nil
```

Comment:
This is what I wanted to avoid with the struct that enforces oneof Paths, PathHashPrefixes.

I'm more comfortable with the second approach https://github.com/ethan-lowman-dd/go-tuf/commit/8eac49ac1dfc789722d85f5c53731a07a89e012d
We are bubbling complexity:
- a user can create an invalid DelegatedRole with the struct
- we have to add error handling everywhere for no real error (can be checked at encoding and decoding)

Reply:
Unless the type system can fully enforce all validation (which it can't for more complex kinds of validation), I don't see a way to avoid storing invalid data in a struct. In my opinion, matching the type definition directly to the spec is much easier to understand (for library authors and users alike), and validating before use and when marshalling/unmarshalling provides the necessary properties, matching the spec as close as possible.

Using a boolean flag (`MatchWithHashPrefixes`) and field with two possible meanings (`PathMatchers`) to implement the "one of" constraint forces users to understand quirks of the library's implementation, rather than following the spec. 

Additionally, the boolean field is not extensible. If the TUF spec introduces a third field that is also subject to the "one of" constraint, the boolean flag would no longer properly implement the constraint across the three fields.